Created new information_schema.rocksdb_deadlock table.

george-lorch opened the pull request:
This table tracks historic deadlocks based on the functionality introduced in
"Display Deadlock Paths".
Copied the deadlock_tracking test and touched it up to test the new table as
opposed to the SHOW ENGINE ROCKSDB TRANSACTION STATUS.

Commit:
Created new information_schema.rocksdb_deadlock table.
- This table tracks historic deadlocks based on the functionality introduced in
  "Display Deadlock Paths".
- Factored existing functionality for SHOW ENGINE ROCKSDB TRANSACTION STATUS
  for historic deadlock into structure similar to trx_info for consistent use
  across SERTS and I_S.DEADLOCK.
- Cleaned up some value naming semantics for SERTS result, added ROLLED BACK
  field that identifies the 'victim' transaction, and removed redundant
  transaction output that identifies the victim.
- Copied the deadlock_tracking test and touched it up to test the new table as
  opposed to the SHOW ENGINE ROCKSDB TRANSACTION STATUS.
- Touched up rocksdb.deadlock_tracking test.

lth commented:
The tests `main.mysqlshow`, `main.mysqld--help-notwin-profiling`, `main.mysqld--help-notwin` need updating.

george-lorch commented:
Ahh yeah, sorry about that. I forgot that in fb-mysql I_S, sysvars and others show in these tests whereas they do not in PS. I'll get them updated and re-push

lth commented:
It would also be great if the existing SHOW ENGINE code could reuse this code, but it's okay if you don't want to do it.

george-lorch commented:
That is not a problem at all. I can remove the 'parallel' implementation bits in the Rdb_snapshot_status that fills the string data buffer and change the populate_deadlock_buffer to behave more like the loop in the I_S fill or even remove it entirely and put the loop right in populate_deadlock_buffer. I considered that initially but figured I would keep things apart and see what kind of review feedback emerged.

If you have any suggestions on the schema to make it more clear of the relationships between rows for an individual deadlock or any other info that might be useful, please let me know.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3367,8 +3347,52 @@ class Rdb_snapshot_status : public Rdb_tx_list_walker {
     for (auto path_entry : dlock_buffer) {
-      m_data += get_dlock_path_info(path_entry);
+      std::string path_data;
+      const auto dl_info = get_dl_path_trx_info(path_entry);
```
I think `get_dl_path_trx_info` does not need to be called if limit is exceeded.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3367,8 +3347,52 @@ class Rdb_snapshot_status : public Rdb_tx_list_walker {
+        for (auto it = dl_info.path.begin(); it != dl_info.path.end(); it++) {
+          const auto trx_info = *it;
+          path_data += format_string(
+              "TRANSACTIONID: %u\n"
```
This isn't from your diff, but I think "TRANSACTION ID" would be more consistent.

lth commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3367,8 +3347,52 @@ class Rdb_snapshot_status : public Rdb_tx_list_walker {
+          /* prints the txn id of the transaction that caused the deadlock */
+          auto deadlocking_txn = *(path_entry.path.end() - 1);
+          path_data += format_string(
+              "\n--------TRANSACTIONID: %u GOT DEADLOCK---------\n",
```
Same thing, "TRANSACTION ID".

lth commented on storage/rocksdb/rdb_i_s.cc:
```diff
@@ -1476,6 +1476,117 @@ static int rdb_i_s_trx_info_init(void *const p) {
+    ROCKSDB_FIELD_INFO("WAITING_KEY", FN_REFLEN + 1, MYSQL_TYPE_STRING, 0),
+    ROCKSDB_FIELD_INFO("LOCK_TYPE", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+    ROCKSDB_FIELD_INFO("INDEX NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
+    ROCKSDB_FIELD_INFO("TABLE NAME", NAME_LEN + 1, MYSQL_TYPE_STRING, 0),
```
Can we standardize on underscores vs spaces? I think we tend to use underscore.

lth commented on storage/rocksdb/ha_rocksdb.h:
```diff
@@ -91,6 +91,25 @@ struct Rdb_trx_info {
+    std::string table_name;
+  };
+  std::vector <Rdb_dl_trx_info> path;
+  ulonglong causer_trx_id;
```
I think the intention of the message `TXN_ID GOT DEADLOCK` was to denote which transaction got rolled back. It's not so much the causer as it is the victim (although in rocksdb, the causer is *always* the victim).

How about a bool column called "Error" or "Rolled back" or something to denote that transaction got deadlock error? It would avoid repeating the id for every row.

george-lorch replied:
That actually makes a lot more sense than re-listing the transaction. Perhaps something similar can be added to the txn block output in the SHOW STATUS.

lth reviewed (commented):
I think the schema looks pretty good overall!

Can you also rebase?

george-lorch commented:
All great comments. I am travelling the next two days so may not get a push with the changes until Friday or next week, just didn't want you to think I abandoned it.

george-lorch commented:
OK @lth, I rebased and cleaned up all things as discussed and re-worked the output structure of both the I_S table and the SERTS output. I eliminated the redundant reporting of the victim/rolled back transaction and added the simple ROLLED BACK field which should indicate the same idea. Cleaned up/normalized the field names, etc... Please take another look when you have a few minutes and let me know if I missed or misunderstood anything or if you have any more ideas on how to improve the feature. Thanks!

lth commented on mysql-test/suite/rocksdb/r/deadlock_tracking.result:
```diff
@@ -59,22 +59,15 @@ KEY
-INDEX NAME: PRIMARY
-TABLE NAME: test.t
-
---------TXN_ID GOT DEADLOCK---------
```
Actually, what I was going for was that the SHOW ENGINES code can keep using `TRANSACTION ID GOT DEADLOCK`, since I think that's a more concise way of expressing that info.

I think the "rolled back" column makes more sense in the I_S table, than in the show engines command.

george-lorch replied:
Ahh, OK, no problem, I can bump that back, but what about the original also re-printing the deadlocked txn info ? Maybe just list the transaction path and finish with the TRANSACTION ID <n> GOT DEADLOCK ?

lth replied:
Yeah, we can take out the reprinted transaction.

lth reviewed (commented):
One more test to update: `main.information_schema`.

george-lorch commented:
OK, I added back the 'GOT DEADLOCK' and removed the 'ROLLED BACK' on the status output as discussed, re-recorded the test and ran another mtr cycle.

lth commented:
The diff looks good to me. The only thing is to update `main.information_schema`.

george-lorch commented:
Thanks Manuel, I fixed up the main.information_schema result and just re-pushed.
